File: 57.insert-interval/solution.cpp
/**
 * Definition for an interval.
 * struct Interval {
 *     int start;
 *     int end;
 *     Interval() : start(0), end(0) {}
 *     Interval(int s, int e) : start(s), end(e) {}
 * };
 */
class Solution {
public:
	vector<Interval> insert(vector<Interval>& intervals, Interval newInterval) {
		vector<Interval> result;
		if (intervals.empty())
		{
			result.push_back(newInterval);
			return result;
		}
		vector<Interval>::iterator it1 = lower_bound(intervals.begin(), intervals.end(), newInterval.start, [](const Interval &interval, int i)->bool{return interval.end < i;});
		vector<Interval>::iterator it3 = upper_bound(intervals.begin(), intervals.end(), newInterval.end, [](int i, const Interval &interval)->bool{return interval.start > i;});
		vector<Interval>::iterator it2 = it3 - 1;
		result.insert(result.end(), intervals.begin(), it1);
		result.emplace_back(it1 != intervals.end() && it1->start < newInterval.start ? it1->start : newInterval.start,
			it2->end > newInterval.end ? it2->end : newInterval.end);
		result.insert(result.end(), it3, intervals.end());
		return result;
    }
};

File: 57.insert-interval/solution.go
func insert(intervals [][]int, newInterval []int) [][]int {
	m := make(map[int]int)
	for _, interval := range intervals {
		m[interval[0]]++
		m[interval[1]]--
	}
	m[newInterval[0]]++
	m[newInterval[1]]--
	var keys []int
	for k := range m {
		keys = append(keys, k)
	}
	sort.Ints(keys)
	temp := 0
	var last int
	var result [][]int
	for _, key := range keys {
		if m[key] > 0 {
			if temp == 0 {
				last = key
			}
			temp += m[key]
		} else if m[key] < 0 {
			temp += m[key]
			if temp == 0 {
				result = append(result, []int{last, key})
			}
		} else {
			if temp == 0 {
				result = append(result, []int{key, key})
			}
		}
	}
	return result
}
